src/bufferer.rs

```diff
@@ -1,294 +1,587 @@
+// let value_i16: i16 = read_number!(deserializer, i16)?;
+// let value_i32: i32 = read_number!(deserializer, i32)?;
+// let value_i64: i64 = read_number!(deserializer, i64)?;
+// let value_f32: f32 = read_number!(deserializer, f32)?;
+// let value_f64: f64 = read_number!(deserializer, f64)?;
+#[allow(unused_macros)]
+macro_rules! read_number {
+    ($deserializer:expr, u8) => {
+        $deserializer.read_u8()
+    };
+    ($deserializer:expr, u16) => {
+        $deserializer.read_u16()
+    };
+    ($deserializer:expr, u32) => {
+        $deserializer.read_u32()
+    };
+    ($deserializer:expr, u64) => {
+        $deserializer.read_u64()
+    };
+    ($deserializer:expr, i8) => {
+        $deserializer.read_i8()
+    };
+    ($deserializer:expr, i16) => {
+        $deserializer.read_i16()
+    };
+    ($deserializer:expr, i32) => {
+        $deserializer.read_i32()
+    };
+    ($deserializer:expr, i64) => {
+        $deserializer.read_i64()
+    };
+    ($deserializer:expr, f32) => {
+        $deserializer.read_f32()
+    };
+    ($deserializer:expr, f64) => {
+        $deserializer.read_f64()
+    };
+}
 
-        let value = std::str::from_utf8(sub_buf)
-            .map_err(|_| PacketBad)?
-            .to_string();
+// Macro Usage
+//
+// let utf8_string: String = read_string!(deserializer, utf8, 10)?;
+// let utf16_string: String = read_string!(deserializer, utf16, 5)?;
+//
+// --------------------
+//
+// #[allow(unused_macros)]
+// macro_rules! read_string {
+// ($deserializer:expr, utf8, $length:expr) => {
+// $deserializer.read_utf8_string($length)
+// };
+// ($deserializer:expr, optional_utf8, $length:expr) => {
+// $deserializer.read_optional_utf8_string($length)
+// };
+// ($deserializer:expr, remaining_utf8) => {
+// $deserializer.read_remaining_utf8_string()
+// };
+// ($deserializer:expr, utf16, $length:expr) => {
+// $deserializer.read_utf16_string($length)
+// };
+// }
 
-        self.move_position_ahead(value.len());
-        Ok(value)
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use byteorder::{BigEndian, LittleEndian};
+
+    struct TestData {
+        u8_val: u8,
+        u16_val: u16,
+        u32_val: u32,
+        u64_val: u64,
+        i8_val: i8,
+        i16_val: i16,
+        i32_val: i32,
+        i64_val: i64,
+        f32_val: f32,
+        f64_val: f64,
     }
 
-    pub fn get_string_utf16(&mut self) -> GDResult<String> {
-        let sub_buf = self.remaining_data();
-        if sub_buf.is_empty() {
-            return Err(PacketUnderflow);
-        }
+    const TEST_DATA: TestData = TestData {
+        u8_val: 0xD6,
+        u16_val: 0xC00F,
+        u32_val: 0xADBFC0E1,
+        u64_val: 0x47F12D29C7F74E9F,
+        i8_val: -126,
+        i16_val: -24368,
+        i32_val: -1874965832,
+        i64_val: -2049853289529583747,
+        f32_val: -0.6893411,
+        f64_val: 0.24812124701362135,
+    };
+
+    fn get_test_data<T>(f: fn(&TestData) -> T) -> T { f(&TEST_DATA) }
```

Comment:
Why do we have this function? To make swapping test data more easily?

Reply:
The function `get_test_data` returns the result of invoking `f` with a reference to the constant object `TEST_DATA` of type `TestData`. `get_test_data` provides a way to pass a function that operates on `TestData` objects, and then applies that function to a pre-defined `TestData` object. The return value of `get_test_data` will be the value returned by the function `f` when passed `TEST_DATA`. The main purpose of the `get_test_data` function is to provide a standardized way of accessing test data by encapsulating the test data in a constant object, and passing a function to operate on that data, `get_test_data` ensures that the test data is kept in a consistent and immutable state.

Reply:
Its not actually needed but its there for utility use.